compiler/rustc_lexer/src/lib.rs

```diff
@@ -277,6 +277,62 @@ pub fn strip_shebang(input: &str) -> Option<usize> {
     None
 }
 
+/// Frontmatter is a special attribute type reserved for use by external tools
+///
+/// This must be called after [`strip_shebang`]
+pub fn strip_frontmatter(input: &str) -> Option<usize> {
+    // Whitespace may precede a frontmatter but must end with a newline
+    let mut rest = input;
+    while !rest.is_empty() {
+        let without_spaces = rest.trim_start_matches(|c| is_whitespace(c) && c != '\n');
+        let without_nl = without_spaces.trim_start_matches('\n');
+        if without_nl.len() == rest.len() {
+            // nothing trimmed
+            break;
+        } else if without_nl.len() == without_spaces.len() {
+            // either not a frontmatter or invalid opening
+            return None;
+        }
+        rest = without_nl;
+    }
+
+    // Opens with a line that starts with 3+ `-` followed by an optional identifier
+    const FENCE_CHAR: char = '-';
+    let fence_end =
+        rest.char_indices().find_map(|(i, c)| (c != FENCE_CHAR).then_some(i)).unwrap_or(rest.len());
+    if fence_end < 3 {
+        // either not a frontmatter or invalid frontmatter opening
+        return None;
+    }
+    let (fence_pattern, rest) = rest.split_at(fence_end);
+    let (info, rest) = rest.split_once("\n").unwrap_or((rest, ""));
+    let info = info.trim_matches(is_whitespace);
+    if !info.is_empty() && !is_ident(info) {
+        // invalid infostring
+        return None;
+    }
+
+    // Ends with a line that starts with a matching number of `-` only followed by whitespace
+    let rest = if let Some(rest) = rest.strip_prefix(fence_pattern) {
+        rest
+    } else {
+        let nl_fence_pattern = format!("\n{fence_pattern}");
+        let Some(frontmatter_nl) = rest.find(&nl_fence_pattern) else {
+            // frontmatter close is required
+            return None;
+        };
+        &rest[frontmatter_nl + nl_fence_pattern.len()..]
+    };
+
+    let (line, rest) = rest.split_once("\n").unwrap_or((rest, ""));
+    let line = line.trim_matches(is_whitespace);
+    if !line.is_empty() {
+        // invalid close, even if there are extra `-`s
```

Comment:
```suggestion
        // invalid close if there are extra characters beyond the original fence pattern
```

Reply:
`extra '-'`s is the key part I am trying to emphasize because that is a case that can easily be overlooked, by implementers or users.

Reply:
Went with a variant that preserves that information.  Squashed it, see https://github.com/rust-lang/rust/compare/26f461654c1e8ef0d93b5267b3873685688d25c3..2501d454739ef26ef52e213fe5b12e94fde3f48c